# Project: Wrangling and Analyze Data

## Data Gathering
In the cell below, gather **all** three pieces of data for this project and load them in the notebook. **Note:** the methods required to gather each data are different.
1. Directly download the WeRateDogs Twitter archive data (twitter_archive_enhanced.csv)

In [228]:
import csv
import urllib.request
import pandas as pd
import tweepy
import os
import os.path
import requests
import json
import time

In [229]:
cwd = os.getcwd() 
base_path = os.getcwd()
cwd

'C:\\Users\\mikle\\Downloads\\wgu\\DA\\project3'

In [230]:
def file_exists(base_p, file_name):
    return(os.path.exists(os.path.join(base_p, file_name)))

In [231]:
if (not(file_exists(base_path,"twitter-archive-enhanced.csv"))):
    url = 'https://d17h27t6h515a5.cloudfront.net/topher/2017/August/59a4e958_twitter-archive-enhanced/twitter-archive-enhanced.csv'
    file_to_read = urllib.request.urlopen(url)
    df_ta = pd.read_csv(file_to_read)
    df_ta.to_csv(tw_arch_file)
else:
    file_to_read = os.path.join(base_path, "twitter-archive-enhanced.csv")
    df_ta = pd.read_csv(file_to_read)

Unnamed: 0            tweet_id  in_reply_to_status_id  in_reply_to_user_id  \
0           0  892420643555336193                    NaN                  NaN   
1           1  892177421306343426                    NaN                  NaN   
2           2  891815181378084864                    NaN                  NaN   

                   timestamp  \
0  2017-08-01 16:23:56 +0000   
1  2017-08-01 00:17:27 +0000   
2  2017-07-31 00:18:03 +0000   

                                              source  \
0  <a href="http://twitter.com/download/iphone" r...   
1  <a href="http://twitter.com/download/iphone" r...   
2  <a href="http://twitter.com/download/iphone" r...   

                                                text  retweeted_status_id  \
0  This is Phineas. He's a mystical boy. Only eve...                  NaN   
1  This is Tilly. She's just checking pup on you....                  NaN   
2  This is Archie. He is a rare Norwegian Pouncin...                  NaN   

   retweeted_status_user_id retweeted_status_timestamp  \
0                       NaN                        NaN   
1                       NaN                        NaN   
2                       NaN                        NaN   

                                       expanded_urls  rating_numerator  \
0  https://twitter.com/dog_rates/status/892420643...                13   
1  https://twitter.com/dog_rates/status/892177421...                13   
2  https://twitter.com/dog_rates/status/891815181...                12   

   rating_denominator     name doggo floofer pupper puppo  
0                  10  Phineas  None    None   None  None  
1                  10    Tilly  None    None   None  None  
2                  10   Archie  None    None   None  None

2. Use the Requests library to download the tweet image prediction (image_predictions.tsv)

In [232]:
url = 'https://d17h27t6h515a5.cloudfront.net/topher/2017/August/599fd2ad_image-predictions/image-predictions.tsv'
response = urllib.request.urlopen(url)
df_ipr = pd.read_csv(response, sep = '\t')


tweet_id                                          jpg_url  \
0  666020888022790149  https://pbs.twimg.com/media/CT4udn0WwAA0aMy.jpg   
1  666029285002620928  https://pbs.twimg.com/media/CT42GRgUYAA5iDo.jpg   
2  666033412701032449  https://pbs.twimg.com/media/CT4521TWwAEvMyu.jpg   

   img_num                      p1   p1_conf  p1_dog                  p2  \
0        1  Welsh_springer_spaniel  0.465074    True              collie   
1        1                 redbone  0.506826    True  miniature_pinscher   
2        1         German_shepherd  0.596461    True            malinois   

    p2_conf  p2_dog                   p3   p3_conf  p3_dog  
0  0.156665    True    Shetland_sheepdog  0.061428    True  
1  0.074192    True  Rhodesian_ridgeback  0.072010    True  
2  0.138584    True           bloodhound  0.116197    True

get all tweet_ids

In [233]:
tweet_ids = df_ta['tweet_id'].astype(str)
tweet_ids.shape[0]

2356

3. Use the Tweepy library to query additional data via the Twitter API (tweet_json.txt)

In [234]:
path_to_bearer_token = os.path.join(base_path, "bearer_token.txt")
with open(path_to_bearer_token, 'r') as f:    
    bearer_token = f.read()

In [236]:
def get_tweat_url(tw_id):
    twitter_url = "https://api.twitter.com/2/tweets/" + str(tw_id)
    query_params = {#'id': tw_id,
                    'expansions': 'author_id,in_reply_to_user_id,geo.place_id',
                    'tweet.fields': 'id,author_id,in_reply_to_user_id,geo,conversation_id,created_at,lang,public_metrics,referenced_tweets,reply_settings,source,text',
                    'user.fields': 'id,name,username,created_at,description,public_metrics,verified',
                    'place.fields': 'full_name,id,country,country_code,geo,name,place_type',
                    'next_token': {}}
    return (twitter_url, query_params)

def get_tweats_url(tw_ids):
    twitter_url = "https://api.twitter.com/2/tweets"
    query_params = {'ids': tw_ids,
                    'expansions': 'author_id,in_reply_to_user_id,geo.place_id',
                    'tweet.fields': 'id,author_id,in_reply_to_user_id,geo,conversation_id,created_at,lang,public_metrics,referenced_tweets,reply_settings,source,text',
                    'user.fields': 'id,name,username,created_at,description,public_metrics,verified',
                    'place.fields': 'full_name,id,country,country_code,geo,name,place_type',
                    'next_token': {}}
    return (twitter_url, query_params)

def connect_to_twitter(url, headers, params, next_token = None):
    params['next_token'] = next_token
    resp = requests.request("GET", url, headers = headers, params = params)
    print("Endpoint Response Code: " + str(resp.status_code))
    if resp.status_code != 200:
        raise Exception(resp.status_code, resp.text)
    return (resp.json())
  
    
def get_ids_string(start, quantity, tw_ids_series): #format sting with ids for pulling
    result = ''
    for i in range(start, start + quantity):
        if i in tw_ids_series.index:
            if len(result) > 0:
                result = result +','+ tw_ids_series[i]
            else:
                result = tw_ids_series[i]
        else:
            i-=1
            break
    return (result, i-start+1)
def add_timestamp(my_series):
    print('type_my_ser:', type(my_series) )
    cur_time = time.time()
    with open(os.path.join(base_path, 'rqsts.csv'),'a') as file:
        file.write(str(cur_time))
        file.write('\n')
    my_series = my_series.values.tolist()
    my_series.append(cur_time)
    my_series = pd.Series(my_series)
    return(my_series)
    
def make_request(tweets_ids, time_window = 1700, requests_limit = 50):
    if (file_exists(base_path, 'rqsts.csv')):
        rqsts = pd.read_csv(os.path.join(base_path, 'rqsts.csv'), names = ['timestamp'])
        rqsts_s = rqsts['timestamp']
    else:
        rqsts_s = pd.Series([])
    current_time = time.time()
    rqsts_f = rqsts_s[rqsts_s > (current_time - time_window)]
    if (rqsts_f.shape[0] < requests_limit):
        #do request
        url = get_tweats_url(tweets_ids)
        headers = {"Authorization": "Bearer " + bearer_token}
        rqsts_f = rqsts_f.values.tolist()
        rqsts_f.append(current_time)
        rqsts_f = pd.Series(rqsts_f)
        rqsts_f.to_csv(os.path.join(base_path, 'rqsts.csv'))
        json_resp = connect_to_twitter(url[0], headers, url[1])
        print("request completed")
        return(json_resp, True)
    else:
        print('limit of request exceeded: requests= ', rqsts_f.shape[0], 'limit=', requests_limit)
        return(0, False)
    
#read_JSON RABOTAET NEPRVILNO https://github.com/pandas-dev/pandas/issues/20608# (pri chtenii)
def pull_all_data(ids_series, batch_size = 5, time_window = 950, req_limit = 50, sleep_time = 10):
    if (file_exists(base_path, 'tweet_json.txt')):
        #workaround the insidious bug in read_json method of pandas: "https://github.com/pandas-dev/pandas/issues/20608#"
        #some random id's in my dataframe became  a little bit different tnan they should be...
        #I spend one hour to understand that my code is fine, it's pandas bug :-)
        tweet_json = pd.read_json(os.path.join(base_path, 'tweet_json.txt'), lines=True, convert_axes=False, dtype={})
        last_id = tweet_json_df.iloc[-1:]['id'].values[0] # the id of the last tweet in the output file (single value)
        idx = (df_ta.index[df_ta['tweet_id'].astype(str) == last_id])
        start_line = (idx[0].item()) #number of string to start, converted to basic python type
    else:
        start_line = 0
    count = 0
    for current_line in range(start_line, ids_series.shape[0], batch_size):
        ids = get_ids_string(current_line, batch_size, ids_series)
        rsp = ('', False)
        while (rsp[1]!= True):
            time.sleep(1)
            rsp = make_request(ids[0],time_window, req_limit)
            if (rsp[1]!= True): #we exceed the limit
                print('waiting ', sleep_time, ' seconds' )
                time.sleep(sleep_time)
        if (not((rsp[0].get('data')) is None)):
            missed_id_flag = False
            for object_number in range(0,len(rsp[0].get('data'))): #check, maybe ids[1] +1
                if (len(rsp[0].get('data')) < batch_size):
                    missed_id_flag = True
                with open('tweet_json.txt', 'a') as out_f:
                    json.dump(rsp[0].get('data')[object_number],out_f)
                    out_f.write('\n')
                    count +=1
            if (missed_id_flag):
                print('Missed id is here ------------>') #monitoring cases where response has less than batch_size data
    print('records completed =', count)

In [237]:
pull_all_data(df_ta['tweet_id'].astype(str), batch_size = 50, time_window = 905, req_limit = 50, sleep_time = 10)

C:\Users\mikle\anaconda3\envs\anac_3.6\lib\site-packages\ipykernel_launcher.py:58: DeprecationWarning: The default dtype for empty Series will be 'object' instead of 'float64' in a future version. Specify a dtype explicitly to silence this warning.


Endpoint Response Code: 200
request completed
Missed id is here ------------>
Endpoint Response Code: 200
request completed
Missed id is here ------------>
Endpoint Response Code: 200
request completed
Missed id is here ------------>
Endpoint Response Code: 200
request completed
Missed id is here ------------>
Endpoint Response Code: 200
request completed
Missed id is here ------------>
Endpoint Response Code: 200
request completed
Missed id is here ------------>
Endpoint Response Code: 200
request completed
Endpoint Response Code: 200
request completed
Missed id is here ------------>
Endpoint Response Code: 200
request completed
Endpoint Response Code: 200
request completed
Endpoint Response Code: 200
request completed
Missed id is here ------------>
Endpoint Response Code: 200
request completed
Missed id is here ------------>
Endpoint Response Code: 200
request completed
Endpoint Response Code: 200
request completed
Endpoint Response Code: 200
request completed
Endpoint Response Code

In [289]:
# find all missed id's in the first 200 rows:
tweet_json = pd.read_json(os.path.join(base_path, 'tweet_json.txt'), lines=True, convert_axes=False, dtype={})
df2 = tweet_json_df[['id','conversation_id']]
df1 = df_ta[['tweet_id','timestamp']].head(200)
df1['id'] = df_ta['tweet_id'].astype(str)
df = pd.merge(df1, df2, on ='id', how ="left")[pd.merge(df1, df2, on ='id', how ="left")['conversation_id'].isna()]
missed_ids = df['tweet_id'].astype(str).str.cat(sep=',')
print(missed_ids)


888202515573088257,873697596434513921,872668790621863937,872261713294495745,869988702071779329,866816280283807744,861769973181624320,856602993587888130,856330835276025856


In [251]:
rsp = make_request(missed_ids,905, 50)
if (not((rsp[0].get('data')) is None)):
    for object_number in range(0,len(rsp[0].get('data'))): 
        if (len(rsp[0].get('data')) < 9):
            print('HERE------------>')
        with open('bbb.txt', 'a') as out_f:
            json.dump(rsp[0].get('data')[object_number],out_f)
            out_f.write('\n')
else:
    print('data = null, so the tweets do not exist now')
               

Endpoint Response Code: 200
request completed
data = null, so the tweets do not exist now


## Assessing Data
In this section, detect and document at least **eight (8) quality issues and two (2) tidiness issue**. You must use **both** visual assessment
programmatic assessement to assess the data.

**Note:** pay attention to the following key points when you access the data.

* You only want original ratings (no retweets) that have images. Though there are 5000+ tweets in the dataset, not all are dog ratings and some are retweets.
* Assessing and cleaning the entire dataset completely would require a lot of time, and is not necessary to practice and demonstrate your skills in data wrangling. Therefore, the requirements of this project are only to assess and clean at least 8 quality issues and at least 2 tidiness issues in this dataset.
* The fact that the rating numerators are greater than the denominators does not need to be cleaned. This [unique rating system](http://knowyourmeme.com/memes/theyre-good-dogs-brent) is a big part of the popularity of WeRateDogs.
* You do not need to gather the tweets beyond August 1st, 2017. You can, but note that you won't be able to gather the image predictions for these tweets since you don't have access to the algorithm used.



First, let merge all 3 dataframes to one:

In [307]:
df2 = tweet_json_df
df1 = df_ta
df1['id'] = df_ta['tweet_id'].astype(str)
df_mergd1 = pd.merge(df1, df2, on ='id', how ="left")
df_ipr['id'] = df_ipr['tweet_id'].astype(str)
master_df = pd.merge(df_mergd1, df_ipr, on ='id', how ="left")

In [308]:
master_df.head(3)

Unnamed: 0          tweet_id_x  in_reply_to_status_id  \
0           0  892420643555336193                    NaN   
1           1  892177421306343426                    NaN   
2           2  891815181378084864                    NaN   

   in_reply_to_user_id_x                  timestamp  \
0                    NaN  2017-08-01 16:23:56 +0000   
1                    NaN  2017-08-01 00:17:27 +0000   
2                    NaN  2017-07-31 00:18:03 +0000   

                                            source_x  \
0  <a href="http://twitter.com/download/iphone" r...   
1  <a href="http://twitter.com/download/iphone" r...   
2  <a href="http://twitter.com/download/iphone" r...   

                                              text_x  retweeted_status_id  \
0  This is Phineas. He's a mystical boy. Only eve...                  NaN   
1  This is Tilly. She's just checking pup on you....                  NaN   
2  This is Archie. He is a rare Norwegian Pouncin...                  NaN   

   retweeted_status_user_id retweeted_status_timestamp  ... img_num  \
0                       NaN                        NaN  ...     1.0   
1                       NaN                        NaN  ...     1.0   
2                       NaN                        NaN  ...     1.0   

          p1   p1_conf p1_dog        p2   p2_conf p2_dog        p3   p3_conf  \
0     orange  0.097049  False     bagel  0.085851  False    banana  0.076110   
1  Chihuahua  0.323581   True  Pekinese  0.090647   True  papillon  0.068957   
2  Chihuahua  0.716012   True  malamute  0.078253   True    kelpie  0.031379   

  p3_dog  
0  False  
1   True  
2   True  

[3 rows x 42 columns]

In [310]:
master_df.describe()

Unnamed: 0    tweet_id_x  in_reply_to_status_id  \
count  2356.000000  2.356000e+03           7.800000e+01   
mean   1177.500000  7.427716e+17           7.455079e+17   
std     680.262939  6.856705e+16           7.582492e+16   
min       0.000000  6.660209e+17           6.658147e+17   
25%     588.750000  6.783989e+17           6.757419e+17   
50%    1177.500000  7.196279e+17           7.038708e+17   
75%    1766.250000  7.993373e+17           8.257804e+17   
max    2355.000000  8.924206e+17           8.862664e+17   

       in_reply_to_user_id_x  retweeted_status_id  retweeted_status_user_id  \
count           7.800000e+01         1.810000e+02              1.810000e+02   
mean            2.014171e+16         7.720400e+17              1.241698e+16   
std             1.252797e+17         6.236928e+16              9.599254e+16   
min             1.185634e+07         6.661041e+17              7.832140e+05   
25%             3.086374e+08         7.186315e+17              4.196984e+09   
50%             4.196984e+09         7.804657e+17              4.196984e+09   
75%             4.196984e+09         8.203146e+17              4.196984e+09   
max             8.405479e+17         8.874740e+17              7.874618e+17   

       rating_numerator  rating_denominator    tweet_id_y      img_num  \
count       2356.000000         2356.000000  2.075000e+03  2075.000000   
mean          13.126486           10.455433  7.384514e+17     1.203855   
std           45.876648            6.745237  6.785203e+16     0.561875   
min            0.000000            0.000000  6.660209e+17     1.000000   
25%           10.000000           10.000000  6.764835e+17     1.000000   
50%           11.000000           10.000000  7.119988e+17     1.000000   
75%           12.000000           10.000000  7.932034e+17     1.000000   
max         1776.000000          170.000000  8.924206e+17     4.000000   

           p1_conf       p2_conf       p3_conf  
count  2075.000000  2.075000e+03  2.075000e+03  
mean      0.594548  1.345886e-01  6.032417e-02  
std       0.271174  1.006657e-01  5.090593e-02  
min       0.044333  1.011300e-08  1.740170e-10  
25%       0.364412  5.388625e-02  1.622240e-02  
50%       0.588230  1.181810e-01  4.944380e-02  
75%       0.843855  1.955655e-01  9.180755e-02  
max       1.000000  4.880140e-01  2.734190e-01

In [331]:
master_df.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 2356 entries, 0 to 2355
Data columns (total 42 columns):
 #   Column                      Non-Null Count  Dtype              
---  ------                      --------------  -----              
 0   Unnamed: 0                  2356 non-null   int64              
 1   tweet_id_x                  2356 non-null   int64              
 2   in_reply_to_status_id       78 non-null     float64            
 3   in_reply_to_user_id_x       78 non-null     float64            
 4   timestamp                   2356 non-null   object             
 5   source_x                    2356 non-null   object             
 6   text_x                      2356 non-null   object             
 7   retweeted_status_id         181 non-null    float64            
 8   retweeted_status_user_id    181 non-null    float64            
 9   retweeted_status_timestamp  181 non-null    object             
 10  expanded_urls               2297 non-null   object          

In [316]:
master_df[(master_df['rating_numerator'] <10) | (master_df['rating_numerator'] >20)].count()[0]

464

In [319]:
master_df[(master_df['rating_denominator'] != 10)].count()[0]

23

In [324]:
master_df[(master_df['retweeted_status_id'].isna() == False) | (master_df['in_reply_to_status_id'].isna() == False)].count()[0]

259

In [340]:
master_df['text_y'].isna().sum() #5

28

### Quality issues
1. The column rating_numerator has a minimum of 0, whereas the valid minimum is 10 (" Almost always greater than 10"). And Max is over 1700 that is not good too. 

2. The column rating_denominator: it should be 10, but the min is 0, max is 170.

3. The source column has Html tags. For using these URLs anywhere, we should remove Html tags.

4. Some records (with non-null values in columns like 'in_reply_to_status_id' ) are not original tweets as required

5. Some tweets IDs in df_ta were wrong or deleted in Twitter. (see "#find all missed id's"). It caused null values in all columns related to tweet_json_df data frame (text_y, lang, etc.) 

6. Some of the columns have only a few non-null rows. (geo)

7. All columns related to the df_ipr data frame have fewer non-null values than others. It is caused by less quantity of rows. (2075 compared to 2356)

8. Some of the rows has all three non-dog values in p1, p2, p3 dog columns

9. The expanded_urls column has fewer non-null values than it should be (2297 instead of 2356)

In [341]:
print((master_df['public_metrics'].iloc[0]))

{'retweet_count': 7117, 'reply_count': 145, 'like_count': 34245, 'quote_count': 219}


### Tidiness issues
1. Dataset has a lot of redundant columns such as in_reply_to_status_id and in_reply_to_user_id_x, id and tweet_id, and conversation_id, etc. 
2. doggo, floofer, upper, and puppo it's one categorical variable. It should be one column instead of 4.
3. values in public_metrics are dictionaries. It should be converted to several columns like reply_count, like_count, etc

## Cleaning Data
In this section, clean **all** of the issues you documented while assessing. 

**Note:** Make a copy of the original data before cleaning. Cleaning includes merging individual pieces of data according to the rules of [tidy data](https://cran.r-project.org/web/packages/tidyr/vignettes/tidy-data.html). The result should be a high-quality and tidy master pandas DataFrame (or DataFrames, if appropriate).

In [ ]:
# Make copies of original pieces of data


### Issue #1:

#### Define:

#### Code

In [292]:
tweet_json.drop(['conversation_id', 'created_at', 'text', 'reply_settings', 'lang'
                               , 'source', 'geo'], axis=1, inplace=True)

In [293]:
tweet_json = tweet_json.drop('public_metrics', axis=1).join(pd.DataFrame(tweet_json2.public_metrics.values.tolist()))

In [288]:
print(tweet_json[tweet_json['in_reply_to_user_id'].isna() == False].shape[0])

78


In [294]:
print(tweet_json.head(2))

    author_id                  id referenced_tweets in_reply_to_user_id  \
0  4196983835  892420643555336193               NaN                 NaN   
1  4196983835  892177421306343426               NaN                 NaN   

   retweet_count  reply_count  like_count  quote_count  
0           7117          145       34245          219  
1           5359          172       29687          274  


#### Test

### Issue #2:

#### Define

#### Code

#### Test

## Storing Data
Save gathered, assessed, and cleaned master dataset to a CSV file named "twitter_archive_master.csv".

## Analyzing and Visualizing Data
In this section, analyze and visualize your wrangled data. You must produce at least **three (3) insights and one (1) visualization.**

### Insights:
1.

2.

3.

### Visualization